### PR TITLE
Turn off full solution analysis by default for C#

### PR DESCRIPTION
**Pull request by mavasani**

### CODE REVIEW

**Comment by mavasani** [comment]
retest this please

**Comment by mavasani** [comment]
@srivatsn @heejaechang @jasonmalinowski  This is the port for https://github.com/dotnet/roslyn/pull/10828 to microupdate branch.

File: src/VisualStudio/VisualBasic/Impl/Options/VisualBasicSettingsManagerOptionSerializer.vb [hunk 1]
```diff
@@ -118,6 +118,9 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Options
                     Case FeatureOnOffOptions.FormatOnPaste.Name
                         Return Nothing
                 End Select
+            ElseIf key.Feature = ServiceFeatureOnOffOptions.OptionName Then
+                ' ClosedFileDiagnostics has been deprecated in favor of BasicClosedFileDiagnostics
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualBasic/Impl/Options/VisualBasicSettingsManagerOptionSerializer.vb]: Why are we changing the VB side here? I thought that was staying as is?

File: src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs [hunk 2]
```diff
@@ -36,10 +36,25 @@ public int BringUpOnIdentifier
             set { SetBooleanOption(CompletionOptions.TriggerOnTypingLetters, value); }
         }
 
+        // This SettingStore option has now been deprecated in favor of CSharpClosedFileDiagnostics.
```
Comment (jasonmalinowski) [inline src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs]: XML doc comment, and consider an actual attribute here.

File: src/Features/Core/Portable/Shared/Options/ServiceFeatureOnOffOptions.cs [hunk 3]
```diff
@@ -8,11 +8,33 @@ internal static class ServiceFeatureOnOffOptions
...
+        private const bool CSharpClosedFileDiagnosticsEnabledByDefault = false;
+        private const bool DefaultClosedFileDiagnosticsEnabledByDefault = true;
+
         /// <summary>
         /// this option is solely for performance. don't confused by option name. 
         /// this option doesn't mean we will show all diagnostics that belong to opened files when turned off,
         /// rather it means we will only show diagnostics that are cheap to calculate for small scope such as opened files.
         /// </summary>
-        public static readonly PerLanguageOption<bool> ClosedFileDiagnostic = new PerLanguageOption<bool>(OptionName, "Closed File Diagnostic", defaultValue: true);
+        public static readonly PerLanguageOption<int> ClosedFileDiagnostic = new PerLanguageOption<int>(OptionName, "Closed File Diagnostic", defaultValue: -1);
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/Shared/Options/ServiceFeatureOnOffOptions.cs]: I don't get why this is changing to have the integers like this, especially since we're leaking this integers through the AutomationObject as far as I can tell. Minimally I'd expect this to be a nullable Boolean, with the appropriate conversions happening at the AutomationObject level.

File: src/VisualStudio/CSharp/Impl/Options/CSharpSettingsManagerOptionSerializer.cs [hunk 4]
```diff
@@ -103,6 +103,18 @@ private bool ShouldIncludeOnOffOption(FieldInfo fieldInfo)
 
         protected override string SettingStorageRoot { get { return "TextEditor.CSharp.Specific."; } }
 
+        protected override string GetStorageKeyForOption(IOption option)
+        {
+            var name = option.Name;
+            if (option == ServiceFeatureOnOffOptions.ClosedFileDiagnostic)
+            {
+                // ClosedFileDiagnostics has been deprecated in favor of CSharpClosedFileDiagnostics
```
Comment (jasonmalinowski) [inline src/VisualStudio/CSharp/Impl/Options/CSharpSettingsManagerOptionSerializer.cs]: I'd expect a comment here explaining the whole "why?" that this is trying to fix.

File: src/EditorFeatures/Test2/Diagnostics/DiagnosticServiceTests.vb [hunk 5]
```diff
@@ -1915,6 +1916,9 @@ class MyClass
                        </Workspace>
 
             Using workspace = TestWorkspace.CreateWorkspace(test)
+                ' set csharp closed diagnostic option on
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Test2/Diagnostics/DiagnosticServiceTests.vb]: That's what the next line is doing, yes. Comment seems unnecessary.

File: src/VisualStudio/VisualBasic/Impl/Options/AutomationObject.vb [hunk 6]
```diff
@@ -44,12 +44,26 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Options
...
-                SetBooleanOption(ServiceFeatureOnOffOptions.ClosedFileDiagnostic, value)
+                ' Even though this option has been deprecated, we want to respect the setting if the user has explicitly turned off closed file diagnostics (which is the non-default value for 'ClosedFileDiagnostics').
+                ' So, we invoke the setter only for value = False.
+                If Not value Then
+                    SetIntegerOption(ServiceFeatureOnOffOptions.ClosedFileDiagnostic, value:=0)
+                End If
+            End Set
+        End Property
+
+        Public Property BasicClosedFileDiagnostics As Integer
```
Comment (jasonmalinowski) [inline src/VisualStudio/VisualBasic/Impl/Options/AutomationObject.vb]: My understanding of the plan here was the VB side was untouched.

File: src/Features/Core/Portable/Shared/Options/ServiceFeatureOnOffOptions.cs (hunk 3 above)
Comment (mavasani) [inline src/Features/Core/Portable/Shared/Options/ServiceFeatureOnOffOptions.cs]: We have identified that having a boolean state doesn't suffice for this option anymore. Reasons:

1. We want to track explicitly when the check box setting is untouched by the user versus turned off by unchecking - it is very likely we will turn this option back on by default for some scope of background analysis, and we don't want to break the users who explicitly turned it off, but also want the users who didn't ever notice this option to get back the background analysis.
2. We are considering adding more states for background analysis - e.g. run for current documents, open documents, current project, entire solution (just like the error list). We will need more state values for this option.

To summarize, this is a true 3-state option now, likely to become a 3+ multistate option in future. This change deprecates the 2-state option for the newer version, which needs the integral values to be emitted into the settings.

File: src/VisualStudio/VisualBasic/Impl/Options/AutomationObject.vb (hunk 6 above)
Comment (mavasani) [inline src/VisualStudio/VisualBasic/Impl/Options/AutomationObject.vb]: Yes, if the existing bool option would have sufficed - we are now moving to a multistate integral value for this option and hence deprecating the old option. Changes to VB side doesn't change the end user experience.

**Comment by mavasani** [comment]
test vsi please

**Comment by mavasani** [comment]
@jasonmalinowski I have addressed your PR feedback.

**Comment by jasonmalinowski** [comment]
:+1:

**Comment by mavasani** [comment]
Tests seem to be failing due to unrelated issues. Merging this in to unblock signed build - tests will be tracked by the PRs on master.
